scala-xml/src/main/scala/org/http4s/scalaxml/ElemInstances.scala

```diff
@@ -52,18 +52,23 @@ trait ElemInstances {
   implicit def xml[F[_]](implicit F: Concurrent[F]): EntityDecoder[F, Elem] = {
     import EntityDecoder._
     decodeBy(MediaType.text.xml, MediaType.text.html, MediaType.application.xml) { msg =>
-      val source = new InputSource()
-      msg.charset.foreach(cs => source.setEncoding(cs.nioCharset.name))
-
-      collectBinary(msg).flatMap[DecodeFailure, Elem] { chunk =>
-        source.setByteStream(new ByteArrayInputStream(chunk.toArray))
-        val saxParser = saxFactory.newSAXParser()
-        try DecodeResult.successT[F, Elem](XML.loadXML(source, saxParser))
-        catch {
-          case e: SAXParseException =>
-            DecodeResult.failureT(MalformedMessageBodyFailure("Invalid XML", Some(e)))
-          case NonFatal(e) => DecodeResult(F.raiseError[Either[DecodeFailure, Elem]](e))
-        }
+      DecodeResult {
+        msg.bodyText
+          .through(fs2.data.xml.events())
+          .through(fs2.data.xml.dom.documents)
+          .head
+          .compile
+          .lastOrError
+          .map {
+            _.docElem match {
+              case e: Elem => Right(e)
+              case _ => Left(MalformedMessageBodyFailure("Invalid XML"))
```

Comment:
Instead of this charade, maybe this should become an `EntityEncoder[F, Document]` instead?